Upgrade to Ruby 3.1

tenderlove opened the pull request:
This is basically the same as #21061, but trying to keep the diff smaller.  Tests pass locally with Ruby 3.1:

```
Finished in 5 minutes 29 seconds (files took 5.57 seconds to load)
3649 examples, 0 failures, 26 pending
```

matthewford committed:
Add 3.1 to the build matrix

matthewford committed:
Upgrade ruby circleci orb to 2.0.0

update default ruby version to 3.1

matthewford committed:
Update .ruby-version

Use Ruby 3.1 in the Vargant file

matthewford committed:
Update Dockerfile to use Ruby 3.1

matthewford committed:
Use a branch of rails-settings-cached for Psych 4

https://github.com/mastodon/mastodon/pull/21061#issuecomment-1341779139

matthewford committed:
bump ci cache key

matthewford committed:
kt-paperclip with Psych 4 fix

Specs were failing due to paperclips' do_not_validate_attachment_file_type option not working properly

https://github.com/thoughtbot/paperclip/issues/2195

Paperclip: validate_media_type: false also broke the specs as it still expects a filename validation.

So adding in a validation to appease paperclips requirements.

matthewford committed:
CircleCI missing imagemagic packages

Trying to fix convert issues:
  convert-im6.q16: unable to open image `{:output=': No such file or directory @ error/blob.c/OpenBlob/2924.
  convert-im6.q16: no decode delegate for this image format `' @ error/constitute.c/ReadImage/575.
  convert-im6.q16: unable to open image `q:a=': No such file or directory @ error/blob.c/OpenBlob/2924.
  convert-im6.q16: unable to open image `q:a=': No such file or directory @ error/blob.c/OpenBlob/2924.
  convert-im6.q16: no decode delegate for this image format `' @ error/constitute.c/ReadImage/575.
  ./lib/paperclip/lazy_thumbnail.rb:17:in `make'

matthewford committed:
Run check-i18n on ubuntu-22.04

To match CI

shleeable commented on Dockerfile:
```diff
@@ -2,7 +2,7 @@
 ARG NODE_VERSION="16.18.1-bullseye-slim"
 
-FROM ghcr.io/moritzheiber/ruby-jemalloc:3.0.4-slim as ruby
+FROM ghcr.io/moritzheiber/ruby-jemalloc:3.1.2-slim as ruby
```
3.1.3

matthewford commented:
Thanks @tenderlove I didn't get around to rebasing my branch so happy to close the other branch if this is passing.
That said I do think the Gemfile (refactoring the gemfile was the main pain with the conflicts) could do with the a refactor at some pint to document what the gems do.

tenderlove commented:
> That said I do think the Gemfile (refactoring the gemfile was the main pain with the conflicts) could do with the a refactor at some pint to document what the gems do.

Ya, I totally agree.  Just want to get the Ruby version upgraded, then we can tackle cleaning up the Gemfile

nightpool commented on Gemfile:
```diff
@@ -72,7 +72,7 @@ gem 'premailer-rails'
 gem 'rack-cors', '~> 1.1', require: 'rack/cors'
 gem 'rails-i18n', '~> 6.0'
-gem 'rails-settings-cached', '~> 0.6'
+gem 'rails-settings-cached', '~> 0.6', git: 'https://github.com/zunda/rails-settings-cached.git', branch: 'v0.6.6-aliases-true'
```
I'm overall pretty worried about our dependency on rails-settings-cached considering that it's been an issue with almost every version upgrade. However, the newer version has major breaking changes and no clear way to migrate between versions. I think the data storage used by the most recent version (serialized hash or json data in a column on the accounts table) makes the most sense long-term, but at that point I'm not even sure what we're using a gem for lol. And certainly we'd still need something to handle the global site settings, which don't have an associated model they can live with.

tenderlove replied:
I studied the gem and didn't find a good way to upgrade either.  Then I looked in to what the gem does exactly.  It looks like it's storing app settings in the database, but using a YAML file for app defaults.  All values (from the database or YAML file) are also "cached" inside Redis too.

I was going to try changing the code that uses this to just use an in-memory cache because I couldn't understand the reason for putting things in Redis, but then I _think_ I stumbled on the use-case which is: when settings are updated the changes should be reflected across all web processes without restarting the app server.

I don't know if fetching these settings is a bottleneck, but it could be that just fetching them from Postgres every time would be fast enough.  Then we could just eliminate the gem.  I am not familiar with scaling Mastodon servers or what bottlenecks there are in the app, so I didn't want to just make that call.

nightpool replied:
Agreed, and definitely out of scope for this PR, just something I think this raises that we should think about eventually.

nightpool reviewed (approved):
Changes LGTM but we're going to have to do something about rails-settings-cached eventually

nightpool commented:
@tenderlove I forked rails-settings-cached into the Mastodon organization here, in case we need to update it again in the future. Branch names and commit hashes should be the same, can you the gemfile to point there? https://github.com/mastodon/rails-settings-cached

tenderlove commented:
> @tenderlove I forked rails-settings-cached into the Mastodon organization here, in case we need to update it again in the future. Branch names and commit hashes should be the same, can you the gemfile to point there? https://github.com/mastodon/rails-settings-cached

Yes, I will do that!

tenderlove commented on .circleci/config.yml:
```diff
@@ -95,7 +95,7 @@ jobs:
       - install-system-dependencies
       - run:
-          command: sudo apt-get install -y ffmpeg imagemagick libpam-dev
+          command: sudo apt-get install -y ffmpeg imagemagick libmagickcore-dev libmagickwand-dev libjpeg-dev libpng-dev libtiff-dev libwebp-dev libpam-dev
```
I don't actually know, but I'll try removing it and see what it does to the CI.  I pulled those from the original PR because it seemed important, but I'll verify if they actually are. 😃

tenderlove replied:
@nightpool looks like CI fails without this change (see the commits below) 😅

tenderlove committed:
Revert "CircleCI missing imagemagic packages"

This reverts commit a5b6bed50e151f4c9d90222b152d72409e94b629.

tenderlove committed:
Use Mastodon fork of Rails Settings Cached

The Mastodon fork has changes that allow the gem to work on Ruby 3.1

tenderlove committed:
bump to 3.1.3

tenderlove committed:
Add missing dependencies

The 2.7.x series of the Mail gem depend on net-imap and net-pop, but it
doesn't declare the dependencies.  We need to manually declare the
dependency or the application won't boot on Ruby 3.1 and up

tenderlove committed:
Revert "Revert "CircleCI missing imagemagic packages""

This reverts commit 7baa6c53fcd752fa7e8dfc4f1bfbb033aaa32896.

tenderlove commented:
If there's anything I need to do to push this PR over the line (like squashing commits), please let me know!

shleeable commented:
Thank you very much for this... I'd be interested to see tests using both 3.1.3-ubuntu-20.04 and 3.1.3-ubuntu-22.04

inductor commented:
@tenderlove Hi! I think you might need a rebase? :)

matthewford commented:
It would be good if someone could review this quicker; after a month there are bound to be conflicts.
